File: packages/engine/lib/execution/src/runner/javascript/runner.js
// noinspection BadExpressionStatementJS
import { arrow } from "./lib/execution/src/runner/javascript/apache-arrow-bundle.js";
import { Batches } from "./lib/execution/src/runner/javascript/batch.js";
import {
  ExperimentContext,
  SimInitContext,
  gen_sim_ctx,
} from "./lib/execution/src/runner/javascript/context.js";
import { gen_group_state } from "./lib/execution/src/runner/javascript/state.js";

const make_hash_set = (fields) => {
  const set = Object.create(null);
  for (var i_field = 0; i_field < fields.length; ++i) {
    set[fields[i_field]] = true;
  }
  return set;
};

export function start_experiment(datasets, pkg_init_msgs, pkg_fns) {
  this.batches = new Batches();
  for (var dataset_name in datasets)
    datasets[dataset_name] = JSON.parse(datasets[dataset_name]);
  this.experiment_ctx = new ExperimentContext(datasets);
  this.sims = {};

  this.pkgs = {};
  for (var i_pkg = 0; i_pkg < pkg_init_msgs.length; ++i_pkg) {
    const msg = JSON.parse(pkg_init_msgs[i_pkg]);
    const pkg_owned_fields = []; // TODO: Get from `start_experiment` fn args.

    if (this.pkgs[msg.id]) {
      throw new Error(
        "Duplicate package (id, name): " + msg.id + ", " + msg.name,
      );
    }

    const fns = pkg_fns[i_pkg];
    const pkg_start_experiment = fns[0];

    const pkg = (this.pkgs[msg.id] = {
      name: msg.name,
      type: msg.type,
      owns_field: make_hash_set(pkg_owned_fields),

      start_experiment: pkg_start_experiment,
      start_sim: fns[1],
      run_task: fns[2],

      /// `experiment` argument to package functions, initialized as empty object.
      experiment: {},

      /// `sim` arguments to sim-level (start_sim, run_task) package functions,
      /// initialized as empty objects.
      sims: {},
    });
    if (pkg_start_experiment) {
      // TODO: Get warnings from returned object, if any.
      pkg_start_experiment(pkg.experiment, msg.payload, this.experiment_ctx);
    }
  }
}

/// Each package should have properties `name`, `loaders`, `getters` and `owns_field`.
const maybe_add_custom_fns = (custom_fns, to_add, custom_property, pkg) => {
  to_add = to_add[custom_property];
  if (!to_add) return;

  for (var field_name in to_add) {
    // TODO: Uncomment after propagating owned_fields:
    // if (!pkg.owns_field[field_name]) {
    //     throw new Error(
    //         "Packages can only specify " + custom_property + " for fields they own, not '" +
    //         field_name + "' in " + pkg.name
    //     );
    // }
    if (custom_fns[field_name]) {
      throw new Error(
        "Duplicate '" + field_name + "' in " + pkg.name + " " + custom_property,
      );
    }
    custom_fns[field_name] = to_add[field_name];
  }
};

const load_schema = (bytes) => {
  const reader = new arrow.MessageReader(bytes);
  const schema = reader.readSchema();
  return schema;
};

export function start_sim(
  sim_id,
  agent_schema_bytes,
  msg_schema_bytes,
  ctx_schema_bytes,
  pkg_ids,
  pkg_msgs,
  globals,
) {
  globals = JSON.parse(globals);
  const agent_schema = load_schema(agent_schema_bytes);
  const sim = (this.sims[sim_id] = {
    schema: {
      agent: agent_schema,
      msg: load_schema(msg_schema_bytes),
      ctx: load_schema(ctx_schema_bytes),
    },

    state: [],

    // Context loaders and getters are for columns in the context batch.
    context_loaders: {},
    context_getters: {},

    // State loaders and getters are for columns in state agent batches (not message batches). Due to the
    // possibility of dynamic access to state, state objects need access to loaders, but context objects don't.
    state_loaders: {},
    state_getters: {},

    // ctx initialized below
    // GroupState initialized below
  });

  const init_ctx = new SimInitContext(
    this.experiment_ctx,
    globals,
    agent_schema,
  );
  for (var i_pkg = 0; i_pkg < pkg_ids.length; ++i_pkg) {
    const msg = JSON.parse(pkg_msgs[i_pkg]);

    const pkg = this.pkgs[pkg_ids[i_pkg]];
    const pkg_sim = (pkg.sims[sim_id] = {});

    const pkg_start_sim = pkg.start_sim;
    if (pkg_start_sim) {
      // The return value of the `start_sim` method of the package.js impl
      const pkg_loaders_and_getters = pkg_start_sim(
        pkg.experiment,
        pkg_sim,
        msg,
        init_ctx,
      );
      if (
        pkg_loaders_and_getters &&
        (pkg.type === "context" || pkg.type === "state")
      ) {
        // Init and output packages can't specify loaders or getters.
        maybe_add_custom_fns(
          sim[pkg.type + "_loaders"],
          pkg_loaders_and_getters,
          "loaders",
          pkg,
        );
        maybe_add_custom_fns(
          sim[pkg.type + "_getters"],
          pkg_loaders_and_getters,
          "getters",
          pkg,
        );
      }
    }
  }

  // Context getters might vary across simulation runs.
  const SimContext = gen_sim_ctx(sim.schema.ctx, sim.context_getters);
  sim.ctx = new SimContext(this.experiment_ctx, globals);

  // Agent schema and state getters might vary across simulation runs.
  sim.GroupState = gen_group_state(sim.schema.agent, sim.state_getters);
}

export function run_task(sim_id, i_group, pkg_id, task_message) {
  const pkg = this.pkgs[pkg_id];
  const pkg_run_task = pkg.run_task;
  if (!pkg_run_task) {
    throw new Error(
      "Attempt to run package with no `run_task` function: " +
        pkg_id +
        ", " +
        pkg.name,
    );
  }

  var ret;
  const sim = this.sims[sim_id];
  try {
    if (i_group === null || i_group === undefined) {
      ret =
        pkg_run_task(
          pkg.experiment,
          pkg.sims[sim_id],
          JSON.parse(task_message),
          sim.state,
          sim.ctx,
        ) || {};

      // TODO: Move iteration over groups into `sim.flush_changes`
      //       (create this function for sim-level state, not just group-level).
      ret.changes = [];
      for (var j_group = 0; j_group < sim.state.length; ++j_group) {
        ret.changes[j_group] = sim.state[j_group].flush_changes(sim.schema);
      }
    } else {
      const group_ctx = sim.ctx.get_group(i_group);
      ret =
        pkg_run_task(
          pkg.experiment,
          pkg.sims[sim_id],
          JSON.parse(task_message),
          sim.state[i_group],
          group_ctx,
        ) || {};
      ret.changes = sim.state[i_group].flush_changes(sim.schema);
    }
  } catch (e) {
    return {
      // TODO: We should make a stack field on pkg_errors rather than just passing a string
      pkg_error: `
Error: ${e.toString()}
Stack: ${JSON.stringify(e.stack)}
`,
    };
  }
  return ret;
}

/// Invalidates existing `GroupContext` and `AgentContext` objects.
/// (NB: Any `GroupContext` or `AgentContext` objects must be forgotten at
/// the end of a `run_task` call.)
export function ctx_batch_sync(
  sim_id,
  ctx_batch,
  state_group_start_idxs,
  current_step,
) {
  const sim = this.sims[sim_id];

  ctx_batch = this.batches.sync(ctx_batch, sim.schema.ctx);
  ctx_batch.load_missing_cols(sim.schema.ctx, sim.context_loaders);

  sim.ctx.set_batch(ctx_batch, state_group_start_idxs, current_step);
}

const _sync_pools = (sim, batches, agent_pool, message_pool) => {
  for (var i_group = 0; i_group < agent_pool.length; ++i_group) {
    agent_pool[i_group] = batches.sync(agent_pool[i_group], sim.schema.agent);
    agent_pool[i_group].load_missing_cols(sim.schema.agent, sim.state_loaders);

    message_pool[i_group] = batches.sync(message_pool[i_group], sim.schema.msg);
    message_pool[i_group].load_missing_cols(sim.schema.msg, {});
  }
};

export function state_sync(sim_id, agent_pool, message_pool) {
  const sim = this.sims[sim_id];
  _sync_pools(sim, this.batches, agent_pool, message_pool);

  // Update existing groups and create new groups if the number of groups has increased.
  const state = sim.state;
  for (var i_group = 0; i_group < agent_pool.length; ++i_group) {
    if (state[i_group]) {
      state[i_group].set_batches(agent_pool[i_group], message_pool[i_group]);
    } else {
      state[i_group] = new sim.GroupState(
        agent_pool[i_group],
        message_pool[i_group],
        sim.state_loaders,
      );
    }
  }

  // Delete old groups if the number of groups has decreased.
  for (var i = state.length - 1; i >= agent_pool.length; --i) {
    state[i] = undefined;
  }
}

export function state_interim_sync(
  sim_id,
  group_idxs,
  agent_batches,
  message_batches,
) {
  const sim = this.sims[sim_id];
  for (var i = 0; i < group_idxs.length; ++i) {
    const group_state = sim.state[group_idxs[i]];

    const agent_batch = this.batches.sync(agent_batches[i], sim.schema.agent);
    agent_batch.load_missing_cols(sim.schema.agent, sim.state_loaders);

    const msg_batch = this.batches.sync(message_batches[i], sim.schema.msg);
    msg_batch.load_missing_cols(sim.schema.msg, {});

    group_state.set_batches(agent_batch, msg_batch);
  }
}

/// Invalidates existing `GroupContext` and `AgentContext` objects.
export function state_snapshot_sync(sim_id, agent_pool, message_pool) {
  const sim = this.sims[sim_id];
  _sync_pools(sim, this.batches, agent_pool, message_pool);
  sim.ctx.sync_snapshot({
    agent_pool: agent_pool,
    message_pool: message_pool,
  });
}

File: packages/engine/lib/execution/src/runner/python/runner.py
import sys
import time

# TODO: deserialize JSON on the Rust side (don't serialize it in the first
# place, just convert directly to Python objects using PyO3)
import json
import pyarrow as pa

from batch import Batches
from context import ExperimentContext, SimInitContext
from sim import Sim
from util import format_exc_info

"""
Amount of seconds to sleep before freeing resources.
"""
SLEEP_BEFORE_FREE = 2


# We want to catch everything
# pylint: disable=broad-except
class Runner:
    def __init__(self):
        self.batches = Batches()
        # TODO: should self.sims be a list?
        self.sims = {}
        self.pkgs = {}
        self.experiment_ctx = None

    def start_experiment(self, datasets, package_init_msgs, package_functions):
        datasets_builder = {}
        for (key, value) in datasets.items():
            datasets_builder[key] = json.loads(value)
        self.experiment_ctx = ExperimentContext(datasets_builder)

        user_warnings = []
        user_errors = []

        for (msg, fns) in zip(package_init_msgs, package_functions):
            package_start_experiment = fns["start_experiment"]
            pkg = self.pkgs[msg["id"]] = {
                "name": msg["name"],
                "type": msg["type"],
                "owns_field": set(),
                "start_experiment": package_start_experiment,
                "start_sim": fns["start_sim"],
                "run_task": fns["run_task"],
                "experiment": {},
                "sims": {},
            }
            if package_start_experiment:
                payload = json.loads(msg["payload"])
                try:
                    result = package_start_experiment(
                        pkg["experiment"], payload, self.experiment_ctx
                    )
                    warnings = result.get("warnings")
                    if warnings is not None:
                        user_warnings.extend(warnings)
                except Exception:
                    pkg_name = pkg["name"]
                    origin = "experiment init"
                    exc_info = sys.exc_info()
                    error = (
                        f"Package `{pkg_name}` {origin}: {format_exc_info(exc_info)}"
                    )
                    user_errors.push(error)

        return {"user_warnings": user_warnings, "user_errors": user_errors}

    def _free_after_sent(self):
        """
        Give the Rust process time to receive nng messages and then
        kill the runner.
        """
        # TODO: check/fix shared memory allocation
        #   see https://app.asana.com/0/1201461747883418/1201634225076144/f
        time.sleep(SLEEP_BEFORE_FREE)
        self._free()

    def _free(self):
        """Release all resources."""
        self.batches.free()
        # self.messenger = None
        self.pkgs = None
        self.sims = None

    def start_sim(
        self,
        sim_id,
        agent_schema_bytes,
        msg_schema_bytes,
        ctx_schema_bytes,
        package_ids,
        package_messages,
        globals,
    ):
        """
        Registers a new simulation run and executes each package's simulation-level init.
        Context and state packages can optionally return custom Arrow loader and/or getter
        functions, which are later used internally by context and state objects, respectively,
        during task execution. (For example, loaders/getters returned by context packages
        can affect state package task execution, because state packages can use context
        objects. However, context packages can't affect the behavior of state objects.)

        If an error occurs during a package's custom simulation-level init (whether a
        package or user error), the error is sent to the Rust process and the simulation
        run is unregistered (with the expectation that the Rust process will also stop
        the simulation run).

        :param msg: The simulation-level initialization message from the Rust process.
                    Contains globals and the Arrow schema (both of which might differ
                    between simulation runs), and for each package, a custom payload
                    sent by the package's Rust code.
        """
        self.sims[sim_id] = sim = Sim(
            {
                "agent": pa.ipc.read_schema(pa.py_buffer(agent_schema_bytes)),
                "message": pa.ipc.read_schema(pa.py_buffer(msg_schema_bytes)),
                "context": pa.ipc.read_schema(pa.py_buffer(ctx_schema_bytes)),
            },
            self.experiment_ctx,
            globals,
        )
        sim_init_ctx = SimInitContext(
            self.experiment_ctx,
            globals,
            pa.ipc.read_schema(pa.py_buffer(agent_schema_bytes)),
        )

        user_warnings = []
        user_errors = []

        for i, (pkg_id, pkg) in enumerate(self.pkgs.items()):
            pkg["sims"][sim_id] = pkg_sim_data = {}

            if pkg["start_sim"] is not None:
                payload = package_messages[i]

                try:
                    result = pkg["start_sim"](
                        pkg["experiment"], pkg_sim_data, payload, sim_init_ctx
                    )
                    if self._handle_sim_init_result(
                        sim, pkg, result, user_warnings, user_errors
                    ):
                        self.sims.pop(str(sim_id))
                        return {
                            "user_warnings": user_warnings,
                            "user_errors": user_errors,
                        }

                # Have to catch generic exception, because package could throw anything.
                except Exception:
                    exc_info = sys.exc_info()
                    user_errors.push(
                        f"Package `{pkg.name}` sim init: {format_exc_info(exc_info)}"
                    )
                    self.sims.pop(str(sim_id))
                    return {"user_errors": user_errors, "user_warnings": user_warnings}
        
        return {"user_errors": user_errors, "user_warnings": user_warnings}

    def _handle_sim_init_result(self, sim, pkg, result, user_warnings, user_errors):
        """
        :param sim_id: The new simulation run's id
        :param sim: The new simulation run's data
        :param pkg: The package object, with sim-level data
        :param result: What the package's custom sim init returned
        :return: Whether any user errors occurred during this
                 package's sim init
        """
        # Not checking for None as we are currently not consistent with returns from packages
        if not result:  # Package didn't return anything.
            return False

        if pkg["type"] in ("context", "state"):
            sim.maybe_add_custom_fns(result, "loaders", pkg)
            sim.maybe_add_custom_fns(result, "getters", pkg)

        # TODO: Remove duplication with experiment init result handling.
        pkg_name = pkg["name"]
        prefix = f"Package `{pkg_name}` sim init: "
        warnings = result.get("warnings")
        if warnings is not None:
            warnings = [prefix + w for w in warnings]
            user_warnings.extend(warnings)
        errors = result.get("errors")
        if errors is not None:
            errors = [prefix + e for e in errors]
            user_errors.extend(errors)
            return True

        return False

    def run_task(self, sim_id, group_idx, pkg_id, task_msg):
        """
        Execute a package's task in a specific simulation run, optionally for
        a specific group:
            - Look up and execute the package's custom `run_task` function.
            - Handle errors and warnings.
            - If no errors occurred, flush Arrow data and propagate the message
            that the package returned to the next target (e.g. another runner).

        If an error does occur, it is sent to the Rust process and the
        simulation run is unregistered from the runner.

        :param sim_id: The id of the simulation run from which the task was sent
        :param group_idx: If the task should be executed on a single group's data
                          (i.e. single agent state and message batches), this is
                          the group's index. Otherwise, it's `None` -- indicating
                          that the task should be executed for the whole simulation
                          run.
        :param pkg_id: The id of the package from which the task was sent
        :param task_id: A unique id for this instance of this task -- this changes
                        each time the task is sent from the simulation run's main
                        loop to the runners, but stays the same while the task is
                        being sent between the runners, until the task is complete
                        and execution is returned to the simulation main loop.
        :param task_msg: An optional message chosen by the package's Rust code
        """
        task_msg = json.loads(task_msg)

        sim = self.sims[str(sim_id)]
        if group_idx is None:
            state = sim.state
            ctx = sim.context
        else:
            state = sim.state.get_group(group_idx)
            ctx = sim.context.get_group(group_idx)

        user_errors = []
        pkg = self.pkgs[pkg_id]
        try:
            # TODO: Pass `task_id` to package?
            continuation = (
                pkg["run_task"](
                    pkg["experiment"], pkg["sims"][str(sim_id)], task_msg, state, ctx
                )
                or {}
            )
        except Exception:
            # Have to catch generic Exception, because package could throw anything.
            pkg_name = pkg["name"]
            origin = "run_task"
            exc_info = sys.exc_info()
            error = f"Package `{pkg_name}` {origin}: {format_exc_info(exc_info)}"
            user_errors.append(error)
            self.sims.pop(str(sim_id))
            return

        changes = state.flush_changes(sim.schema)
        if group_idx is not None:
            changes["i_group"] = group_idx
            changes = [changes]

        # TODO: OPTIM chaining if `continuation.target == "Python"`
        # NOTE: this should probably be implemented on the Rust side
        ret = {
            "changes": changes,
            "user_warnings": [],
            "user_errors": user_errors,
            **continuation,
        }
        if not "task" in ret:
            ret["task"] = "{}"
        return ret

    def ctx_batch_sync(self, sim_id, ctx_batch, state_group_start_indices, cur_step):
        """
        Load one simulation run's context batch's shared memory segment
        (if necessary) and native columns from Arrow. Also update the
        simulation run's current step.

        :param sim_id: ID of the simulation run whose context batch to sync
        :param ctx_batch: Object describing how to sync the context batch, with
                          the batch's id, batch version and memory version.
        :param cur_step: Current step of the simulation run -- synced along
                         with the batch because it's also part of context
        """
        sim = self.sims[str(sim_id)]

        ctx_batch = self.batches.sync(ctx_batch, sim.schema["context"])
        ctx_batch.load_missing_cols(sim.schema["context"], sim.context_loaders)

        sim.context.set_batch(ctx_batch)
        sim.context.set_step(cur_step)

    def _load_pools(self, sim: Sim, agent_pool, message_pool):
        """
        Load batches corresponding to batch objects in pools,
        mutating both `self.batches` and `agent_pool`/`message_pool`.

        :param sim: Simulation run data, in particular schemas and custom loaders
        :param agent_pool: List of agent batch objects (i.e. objects with the agent schema)
        :param message_pool: List of message batch objects (i.e. objects with the message schema)
        """
        # `group_index` is used for `agent_pool` and `message_pool`. Using `range(len(..))` is more
        # readable as `enumerate(zip(...))`
        # pylint: disable=consider-using-enumerate
        for group_index in range(len(agent_pool)):
            agent_pool[group_index] = self.batches.sync(
                agent_pool[group_index], sim.schema["agent"]
            )
            agent_pool[group_index].load_missing_cols(
                sim.schema["agent"], sim.state_loaders
            )

            message_pool[group_index] = self.batches.sync(
                message_pool[group_index], sim.schema["message"]
            )
            message_pool[group_index].load_missing_cols(
                sim.schema["message"], sim.state_loaders
            )

    def state_sync(self, sim_id, agent_pool, message_pool):
        """
        Load one simulation run's current state's agent pool and message pool.
        Unlike an interim state sync, this can also update the sizes of the
        pools if the number of groups has changed.

        :param sim_id: ID of the simulation run whose state to sync
        :param agent_pool: List of state agent batch objects
        :param message_pool: List of state message (i.e. outbox) batch objects
        """
        sim = self.sims[str(sim_id)]
        self._load_pools(sim, agent_pool, message_pool)
        sim.state.set_pools(agent_pool, message_pool, sim.state_loaders)

    def state_interim_sync(self, sim_id, group_idxs, agent_batches, message_batches):
        """
        Load a subset of one simulation run's state. This can only sync existing
        groups, so it can't change the number of groups.

        :param sim_id: ID of the simulation run whose state to sync
        :param group_idxs: Indices of groups within the simulation run's state
                           to sync
        :param agent_batches: State agent batch objects. The `i`th object must
                              correspond to the `i`th group to sync (i.e.
                              `agent_batches` and `group_idxs` must have the
                              same length), not the `i`th existing group.
        :param message_batches: State message batch objects
        """
        sim = self.sims[str(sim_id)]
        for idx, group_idx in enumerate(group_idxs):
            agent_batch = self.batches.sync(agent_batches[idx], sim.schema["agent"])
            agent_batch.load_missing_cols(sim.schema["agent"], sim.state_loaders)

            msg_batch = self.batches.sync(message_batches[idx], sim.schema["message"])
            msg_batch.load_missing_cols(sim.schema["message"], {})

            group_state = sim.state.get_group(group_idx)
            group_state.set_batches(agent_batch, msg_batch)

    def state_snapshot_sync(self, sim_id, agent_pool, message_pool):
        # TODO: should self.sims be a list?
        sim = self.sims[str(sim_id)]
        self._load_pools(sim, agent_pool, message_pool)
        sim.context.set_snapshot(agent_pool, message_pool)
